### ISSUE TITLE
Folder created in ff is not synced with webapp

### ISSUE THREAD

**Issue by kayceegb** [Offerel/SyncMarks-Extension#20]
Hi,

I discovered this one yesterday trying importing my real BMs in a test new ff profile. 

I'm not not sure if this should go here or in webapp repo. 

Started with empty ff bookmarks (delete button from extension). Then I export this to webapp. At this point it's ok, in webapp, I can see basic ff folders. 

Then I try to create a folder in ff bookmarks and that fails. I have this in logs

    ==> /var/log/bookmark.log <==
    [27-Nov-2020 14:44:16] - [debug] - admin - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin'
    [27-Nov-2020 14:44:16] - [notice] - admin - 127.0.0.1 - Save session to database.
    [27-Nov-2020 14:44:16] - [debug] - admin - 127.0.0.1 - UPDATE `users` SET `userLastLogin`= 1606488256, `sessionID` = 'adok106o1etslootkk69nmrn76', `userOldLogin`=  WHERE `userID` = '1'
    [27-Nov-2020 14:44:16] - [notice] - admin - 127.0.0.1 - Get userdata from the database
    [27-Nov-2020 14:44:16] - [debug] - admin - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin'
    [27-Nov-2020 14:44:17] - [notice] - admin - 127.0.0.1 - Undefined index: url in /var/www/html/bmsync/index.php on line 329
    [27-Nov-2020 14:44:17] - [warn] - admin - 127.0.0.1 - URL is not a valid URL. Exit now.
    
    ==> /var/log/nginx/access.log <==
    127.0.0.1 - admin [27/Nov/2020:15:44:17 +0100] "POST /bmsync/index.php HTTP/1.1" 200 5 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0"
    
    ==> /var/log/bookmark.log <==
    [27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin'
    [27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - Save session to database.
    [27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - UPDATE `users` SET `userLastLogin`= 1606488261, `sessionID` = 'adok106o1etslootkk69nmrn76', `userOldLogin`=  WHERE `userID` = '1'
    [27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - Get userdata from the database
    [27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin'
    [27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - Edit bookmark request, try to find the bookmark first by url...
    [27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - Undefined index: url in /var/www/html/bmsync/index.php on line 741
    [27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - SELECT `bmID`  FROM `bookmarks` WHERE `bmURL` = '' AND `userID` = 1
    [27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - No unique bookmark found, try to find now by title...
    [27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - SELECT `bmID`  FROM `bookmarks` WHERE `bmTitle` = 'ghi' AND `userID` = 1
    [27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - No Unique entry found, chancel operation and send error to client.
    
    ==> /var/log/nginx/access.log <==
    127.0.0.1 - admin [27/Nov/2020:15:44:21 +0100] "POST /bmsync/index.php HTTP/1.1" 200 5 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0"

Tried with extension from amo and from dev branch. 

K.

**Comment by kayceegb** [comment]
Oh and I found that just now. 

When you export once again, folders are exported correctly.

**Comment by Offerel** [comment]
Maybe some request was lost during transfer. We should keep an eye on this.

**Comment by kayceegb** [comment]
Do you have a way to track it ? 

Does that work for you ?

**Comment by Offerel** [comment]
It works here without such problems. I have tested this several times with Firefox, Chrome, Edge and Kiwi on Android.
To track that, we can only compare the JSON content generated by the browser and whats received at the server and later the export from the database. The problem will be, that the browser generated JSON will ever be different in details, from the DB export, since minor things will change. I have no idea if a JSON validator exists, which ignores such minor changes. Only the logical structure should have the same content.

**Comment by kayceegb** [comment]
It really time for me to upgrade. :/

**Comment by kayceegb** [comment]
I'm still "investigating" on this one as it is a blocker one for me. 

In https://github.com/Offerel/SyncMarks-Webapp/issues/26, we can see that a bookmark creation in FF is sent to addon like this
```
Form data |  
-- | --
client | 4b1615a7-6c0d-434c-9b61-0f321b8b405c
caction | addmark
bookmark | {"id":"QtAmWxB1H_eI","url":"https://duckduckgo.com/?q=%22azerty%22&ia=web","title":"\"azerty\"  at  DuckDuckGo","type":"bookmark","folder":"unfiled_____","nfolder":"Other  Bookmarks","added":1606644599023}
```

But here when I create a folder, the request is sent with backslashes everywhere

```
  |  
-- | --
client | "a67ad85d-111a-4bf9-aa66-6085c0d1315c"
caction | "addmark"
bookmark | "{\"id\":\"LQWkQDqewIFH\",\"title\":\"New   Folder\",\"type\":\"folder\",\"folder\":\"unfiled_____\",\"nfolder\":\"Other  Bookmarks\",\"added\":1606678504737}"
```

And in addition, console adds this:

```
Uncaught SyntaxError: JSON.parse: unexpected end of data at line 1 column 1 of the JSON data
    onload moz-extension://012361d3-6c9e-4dcf-b8ee-c1f86f939a06/scripts/background.js:359
background.js:359:25
```    

There really seems to be a problem somewhere but  you say that everythink works for you. Humm...

**Comment by kayceegb** [comment]
Well, I don't know what happened but a retry to create a bookmark have all these backslashes too. 

```
client | "a67ad85d-111a-4bf9-aa66-6085c0d1315c"
-- | --
caction | "addmark"
bookmark | "{\"id\":\"LliC3lG5yKq4\",\"url\":\"https://duckduckgo.com/?q=%22azerty%22&ia=web\",\"title\":\"\\\"azerty\\\"  at  DuckDuckGo\",\"type\":\"bookmark\",\"folder\":\"unfiled_____\",\"nfolder\":\"Other  Bookmarks\",\"added\":1606682441798}"
```

Maybe a change in FF version. So does not seem to be the cause of the problem.

**Comment by Offerel** [comment]
Regarding this logfile:

```
==> /var/log/bookmark.log <==
[27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin'
[27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - Save session to database.
[27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - UPDATE `users` SET `userLastLogin`= 1606488261, `sessionID` = 'adok106o1etslootkk69nmrn76', `userOldLogin`=  WHERE `userID` = '1'
[27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - Get userdata from the database
[27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - SELECT * FROM `users` WHERE `userName`='admin'
[27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - Edit bookmark request, try to find the bookmark first by url...
[27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - Undefined index: url in /var/www/html/bmsync/index.php on line 741
[27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - SELECT `bmID`  FROM `bookmarks` WHERE `bmURL` = '' AND `userID` = 1
[27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - No unique bookmark found, try to find now by title...
[27-Nov-2020 14:44:21] - [debug] - admin - 127.0.0.1 - SELECT `bmID`  FROM `bookmarks` WHERE `bmTitle` = 'ghi' AND `userID` = 1
[27-Nov-2020 14:44:21] - [notice] - admin - 127.0.0.1 - No Unique entry found, chancel operation and send error to client
```
This seems to be an bookmark edit. First it tries to find the bookmark by url. url key is empty (maybe its an folder), thats why this fails and it searches the entry by its title. It cant find an entry with the title 'ghi' and send an error message back to the browser.

**Comment by kayceegb** [comment]
If you look at the title, it's in the case of a folder creation ...

**Comment by Offerel** [comment]
> 
> 
> Well, I don't know what happened but a retry to create a bookmark have all these backslashes too.
> 
> ```
> client | "a67ad85d-111a-4bf9-aa66-6085c0d1315c"
> -- | --
> caction | "addmark"
> bookmark | "{\"id\":\"LliC3lG5yKq4\",\"url\":\"https://duckduckgo.com/?q=%22azerty%22&ia=web\",\"title\":\"\\\"azerty\\\"  at  DuckDuckGo\",\"type\":\"bookmark\",\"folder\":\"unfiled_____\",\"nfolder\":\"Other  Bookmarks\",\"added\":1606682441798}"
> ```
> 
> Maybe a change in FF version. So does not seem to be the cause of the problem.

This looks like some escape characters. Maybe this is from my change earlier today for the addmark switch in WebApp. I reverted the change in commit [186e202](https://github.com/Offerel/SyncMarks-Webapp/commit/186e2022b1071e30d896767ca571d9306e3570ec)

**Comment by kayceegb** [comment]
No need to revert I thing, this is the data send to webapp from addon seen in XHR debug, so it is before webapp can manipulate that. 

And in addition I think I have something...

**Comment by Offerel** [comment]
OK, i see what you mean. But when i validate the string at the backend, the JSON is ok and the escape chars are removed as expected. So in my opinion, the escape chars are fine. I can add bookmarks correctly.

**Comment by Offerel** [comment]
OK, i think i know, why the folder isnt created at the server. If you click "New Folder" in Firefox, the folder is immediately created as "New Folder", instead when you click on OK. If you change in the dialog the name to something other and click ok, the newly created folder is "edited" and this change is send to the backend.

I can find a fix for this situation tomorrow.

**Comment by kayceegb** [comment]
I have it. 

When you create a folder, I don't know why and if it is necessary, but SyncMarks send data 2 times to addon. 
1st: when you open the "New folder dialog" to create folder, the addon sends a create request with default name "New Folder" which is the default in the dialog. 
2nd: when you enter the name you want for your folder (so you replace "New Folder" with your name) and validate, the addon sends a request to rename the folder created before with the new name. 

But,
1st: the webapp is not able to treat the first request
and
2nd: the request to rename the folder, from addon, is incomplete

I was able to bypass the first with that:
```
diff --git a/index.php b/index.php
index efcac1d..48d644b 100644
--- a/index.php
+++ b/index.php
@@ -326,7 +326,8 @@ if(isset($_POST['caction'])) {
        switch($_POST['caction']) {
                case "addmark":
                        $bookmark = json_decode($_POST['bookmark'], true);
-                       $bookmark['url'] = validate_url($bookmark['url']);
+                       if($bookmark['type'] == 'bookmark')
+                               $bookmark['url'] = validate_url($bookmark['url']);
                        $client = filter_var($_POST['client'], FILTER_SANITIZE_STRING);
                        $ctype = getClientType($_SERVER['HTTP_USER_AGENT']);
                        if(strtolower($ctype) != "firefox") $bookmark = cfolderMatching($bookmark);
```

The second part, is not in my game area. The addon sends this request to rename the folder

```
client | "a67ad85d-111a-4bf9-aa66-6085c0d1315c"
-- | --
caction | "editmark"
bookmark | "{\"title\":\"azerti\"}"
```
which is incomplete. The webapp don't which folder rename to, here, azerti. The addon should be fixed for that.

**Comment by kayceegb** [comment]
> OK, i see what you mean. But when i validate the string at the backend, the JSON is ok and the escape chars are removed as expected. So in my opinion, the escape chars are fine. I can add bookmarks correctly.

Bookmarks yes, but not folders. And the issue is about folders ;)

**Comment by Offerel** [comment]
It seems i have found a workaround, which indeed wasn't that easy. The bookmark API will not track information from before a change is changed. It reflects only the changed state. Because of that, specially for a folder, i get only the new title with no information what it was before that change. To make it more worse, it sends this information before someone clicks the ok button (and reverts the changes, when i click chancel, resulting in calling the remove bookmark call).

Luckily i found a way to get the bookmark id, what Firefox the Id is the right one. I can search parts of the missing information with this id in another function, returning the missed information. Because Firefox will change this ID in an attack of senselessness, this ID is not reliably identical to the ID in the backend. 

So i can only try to find the folder which has changed by a handful of vague information, in hope to find a unique match.

I have made a commit for the extension [fe68b9f
](https://github.com/Offerel/SyncMarks-Extension/commit/fe68b9fb9bda66162ebd035feb5b71216b1fc78f) and another commit for the webapp [27c530c
](https://github.com/Offerel/SyncMarks-Webapp/commit/27c530c802a71f76d055ae774981c0c3dfc03cf4). 

There could be odd situations, where the backend cant find a match. In this case it doesn't change anything and send an error message to the extension.

**Comment by kayceegb** [comment]
I don't know if you refer to that when you say "odd", but trying creating a second folder do not works.

**Comment by Offerel** [comment]
Not exactly. Odd means for example if there are changes only at the backend for example, which are not synced to the browser, there can be a situation where the backend couldn't find a unique match for editing and that's why the exit fails, resulting, that the folder isn't created at all at the backend.

A fullsync, meaning a full export or import should solve such a situation. I can't do much against that. Only when the bookmark api accept valid id's and does not change them from itself would solve that completely.

The strange thing is, that FF keeps the id's when using the internal sync mechanism, but it exchanges the id's with the bookmark api. But that's not a bug, it's only the way, the bookmark api is designed accross all browsers. But yes, i hate this.

**Comment by kayceegb** [comment]
Ok. After a full export, creating folder{1,2,3,4, ...} works if you don't create the folder via clicking a bookmark.

P.S.: we have to talk about "import" later lol

**Comment by Offerel** [comment]
at the moment im offline. My dev-server is gone to Valhalla :-( i message you, when i got a new device.

**Comment by kayceegb** [comment]
I'm not sure I understand what you mean by Valhalla, but my supposition is that it's not a good situation for him. :/

I continue testing on my side. I already have 2 or 3 modifications. 

I hope you can be "online" again soon.

**Comment by Offerel** [comment]
He is dead. I fear the sd-card mount on this raspberry pi is defect, so i have to buy a new raspberry pi.

Valhalla in old Vikings myths are the place where you come to, when you have died. You sit there with the gods for some sort of dinner...

**Comment by kayceegb** [comment]
Oh ! If he is with gods, that's fine ^^
Thanks for the clarification. I knew that it was related to vikings without having all the details. 

Are you sure it's not the sd-card ? They can die very quickly sometimes and even more in raspies.

**Comment by Offerel** [comment]
I have tested the card in 2 other devices, where it is ok. Also a new card becomes a corrupted file system after some minutes. Power supply seems to be ok, at least when I boot from network.

But this gives me enough to buy a shiny new version 4 device with more ram.

**Comment by kayceegb** [comment]
> Well, I don't know what happened but a retry to create a bookmark have all these backslashes too.
> 
> ```
> client | "a67ad85d-111a-4bf9-aa66-6085c0d1315c"
> -- | --
> caction | "addmark"
> bookmark | "{\"id\":\"LliC3lG5yKq4\",\"url\":\"https://duckduckgo.com/?q=%22azerty%22&ia=web\",\"title\":\"\\\"azerty\\\"  at  DuckDuckGo\",\"type\":\"bookmark\",\"folder\":\"unfiled_____\",\"nfolder\":\"Other  Bookmarks\",\"added\":1606682441798}"
> ```
> 
> Maybe a change in FF version. So does not seem to be the cause of the problem.

I confirm that the backslashes were due to FF version. Downgrading to latest 68esr from 78esr, XHR debug do not have them.

**Comment by Offerel** [comment]
Now that I have moved my environment to another location, I found a similar problem during testing and created a fix. Unfortunately I saw your mail only later. Since I already have the parentfolder in my http request, my fix looks just a little different.

Commit [3c31b9c](https://github.com/Offerel/SyncMarks-Webapp/commit/3c31b9ccd82bd890ecfd99e7beed7ff35ef20882)

**Comment by kayceegb** [comment]
When I was looking for a solution, I seen the possibility to use IFNULL. I read about compatibility problem with sqlite and even after testing I was unable to make it work it's why I used COALESCE. For sure, it was a syntax issue as with your commit IFNULL is accepted. 

But using your commit I am unable to create a second folder in a folder :/ With mine it is working. Do you see what is missing ?

**Comment by kayceegb** [comment]
> ...If the browser doesn't know the entry yet, it always assigns a new ID, no matter if there is already an ID or not. 
I'm not a developper and even less an addon guru. It's just like a hobby for me. From what I have seen, the browser, at least FF, assigns new ids incrementally, so if both, the browser and the webapp are in sync and you add a new object in webapp, that should replicated correctly, no ?

**Comment by Offerel** [comment]
No. Bookmark ID's in FF are a somewhat unique identifier with 12 random chars and ciphers, like the one i create in my backend. Chromium and derivatives use a single integer. If you add a new bookmark object in FF, it removes the given id (also when it is valid) and creates its own unique id. It does the same when you use in parallel FF-buildin-sync and there is a sync conflict. But it keeps the id or accepts a new id from another client, when there is no conflict.

**Comment by kayceegb** [comment]
Sorry, i was thinking about indexes and actually I forgot completely that there are also IDs. That is ugly this story. 

Subsidiary question, the addon part for folder creation seems ok, so we should leave this one and maybe start a new issue in webapp for the remaining problems. No ?

**Comment by Offerel** [comment]
feel free. I have found the reason why it isnt created, i have only to find a solution.

**Comment by Offerel** [comment]
please see commit [5fc0654](https://github.com/Offerel/SyncMarks-Webapp/commit/5fc0654ac367df0b6695098308c14f88fbbf0c19)
